File: perf/benchmarks/apps/mandelbrot/mandelbrot.cpp
// Mandelbrot — application-style numerical benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

static int64_t mandelbrot(int width, int max_iter) {
    const double inv_w = 2.0 / width;
    int64_t total = 0;
    for (int py = 0; py < width; ++py) {
        const double ci = py * inv_w - 1.0;
        for (int px = 0; px < width; ++px) {
            const double cr = px * inv_w - 1.5;
            double zr = 0.0, zi = 0.0;
            int n = 0;
            while (n < max_iter) {
                const double zr2 = zr * zr;
                const double zi2 = zi * zi;
                if (zr2 + zi2 > 4.0) break;
                zi = 2.0 * zr * zi + ci;
                zr = zr2 - zi2 + cr;
                ++n;
            }
            total += n;
        }
    }
    return total;
}

int main(int argc, char** argv) {
    int width = (argc > 1) ? std::atoi(argv[1]) : 600;
    int max_iter = (argc > 2) ? std::atoi(argv[2]) : 200;

    auto t0 = std::chrono::steady_clock::now();
    int64_t cs = mandelbrot(width, max_iter);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/apps/mandelbrot/mandelbrot.py
# Mandelbrot — application-style numerical benchmark.
#
# Computes a W*W escape-time bitmap of the Mandelbrot set with up to MAX_ITER
# iterations per pixel. Inspired by the Computer Language Benchmarks Game.
# Args: WIDTH MAX_ITER (defaults: 600 200).

import sys
import time


def mandelbrot(width: int, max_iter: int) -> int:
    inv_w = 2.0 / width
    total = 0
    for py in range(width):
        ci = py * inv_w - 1.0
        for px in range(width):
            cr = px * inv_w - 1.5
            zr = 0.0
            zi = 0.0
            n = 0
            while n < max_iter:
                zr2 = zr * zr
                zi2 = zi * zi
                if zr2 + zi2 > 4.0:
                    break
                zi = 2.0 * zr * zi + ci
                zr = zr2 - zi2 + cr
                n += 1
            total += n
    return total


def main() -> None:
    width = int(sys.argv[1]) if len(sys.argv) > 1 else 600
    max_iter = int(sys.argv[2]) if len(sys.argv) > 2 else 200

    t0 = time.perf_counter_ns()
    cs = mandelbrot(width, max_iter)
    elapsed_ns = time.perf_counter_ns() - t0

    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
